File: leetcode-cn/31.下一个排列.py
#
# @lc app=leetcode.cn id=31 lang=python3
#
# [31] 下一个排列
#

# @lc code=start
from typing import List


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        # 从后向前搜索找到第一个小于后一个数的值
        n = len(nums)

        i = n - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1

        if i != -1:
            j = n - 1
            while j >= 0 and nums[j] <= nums[i]:
                j -= 1

            nums[i], nums[j] = nums[j], nums[i]

        nums[i + 1:] = sorted(nums[i + 1:])


# @lc code=end

File: leetcode-cn/C++/31.下一个排列.cpp
/*
 * @lc app=leetcode.cn id=31 lang=cpp
 *
 * [31] 下一个排列
 */

// @lc code=start
#include <algorithm>
#include <vector>

using namespace std;

class Solution
{
public:
    void nextPermutation(vector<int> &nums)
    {
        int n = nums.size();
        if (n == 1)
        {
            return;
        }

        int i = n - 2;

        while (i >= 0)
        {
            if (nums[i] < nums[i + 1])
            {
                break;
            }
            i--;
        }

        if (i < 0)
        {
            sort(nums.begin(), nums.end());
            return;
        }

        int j = n - 1;
        while (nums[j] <= nums[i])
        {
            j--;
        }

        int temp = nums[j];
        nums[j] = nums[i];
        nums[i] = temp;

        sort(nums.begin() + i + 1, nums.end());
    }
};

// @lc code=end
